src/grid_base_ref_c.c

```diff
@@ -0,0 +1,171 @@
+
+#include <stdio.h>
+#include <math.h>
+
+int grid_fill_pol(
+                  const double dr,
+                  const double roffset,
+                  const int lb_cube,
+                  const int lp,
+                  const int cmax,
+                  const double zetp,
+                  double pol[cmax+1][lp+1][2]
+                  ) {
+//
+//   compute the values of all (x-xp)**lp*exp(..)
+//
+//  still requires the old trick:
+//  new trick to avoid to many exps (reuse the result from the previous gridpoint):
+//  exp( -a*(x+d)**2)=exp(-a*x**2)*exp(-2*a*x*d)*exp(-a*d**2)
+//  exp(-2*a*(x+d)*d)=exp(-2*a*x*d)*exp(-2*a*d**2)
+//
+      const double t_exp_1 = exp(-zetp * pow(dr, 2));
+      const double t_exp_2 = pow(t_exp_1, 2);
+
+      double t_exp_min_1 = exp(-zetp * pow(+dr - roffset, 2));
+      double t_exp_min_2 = exp(-2 * zetp * (+dr - roffset) * (-dr));
+      for (int ig=0; ig >= lb_cube; ig--) {
+          const double rpg = ig * dr - roffset;
+          t_exp_min_1 *= t_exp_min_2 * t_exp_1;
+          t_exp_min_2 *= t_exp_2;
+          double pg = t_exp_min_1;
+          // pg  = EXP(-zetp*rpg**2)
+          for (int icoef=0; icoef<=lp; icoef++) {
+              pol[ig+cmax][icoef][0] = pg;
+              pg *= rpg;
+          }
+      }
+
+      double t_exp_plus_1 = exp(-zetp * pow(-roffset,2));
+      double t_exp_plus_2 = exp(-2 * zetp * (-roffset) * (+dr));
+      for (int ig=0; ig >= lb_cube; ig--) {
+          const double rpg = (1-ig) * dr - roffset;
+          t_exp_plus_1 *= t_exp_plus_2 * t_exp_1;
+          t_exp_plus_2 *= t_exp_2;
+          double pg = t_exp_plus_1;
+          // pg  = EXP(-zetp*rpg**2)
+          for (int icoef=0; icoef<=lp; icoef++) {
+              pol[ig+cmax][icoef][1] = pg;
+              pg *= rpg;
+          }
+      }
+
+    return 0;
+}
+
+int grid_collocate_core(
+                        const int grid_size_x,
+                        const int grid_size_y,
+                        const int grid_size_z,
+                        const int grid_lbound_x,
+                        const int grid_lbound_y,
+                        const int grid_lbound_z,
+                        const int lp,
+                        const int cmax,
+                        const double coef_xyz[(lp+1)*(lp+2)*(lp+3)/6],
+                        const double pol[3][cmax+1][lp+1][2],
+                        const int map[3][2*cmax+1],
+                        const int *sphere_bounds,
+                        double grid[grid_size_z][grid_size_y][grid_size_x]
+                        ) {
+
+    int sci = 0;
+
+    const int kgmin = sphere_bounds[sci++];
+    for (int kg=kgmin; kg <= 0; kg++) {
+        const int kg2 = 1 - kg;
+        const int k = map[2][kg + cmax];
+        const int k2 = map[2][kg2 + cmax];
+
+        // initialize coef_xy
+        const int n_coef_xy = (lp+1)*(lp+2)/2;
+        double coef_xy[n_coef_xy][2];
+        for (int i=0; i < n_coef_xy; i++) {
```

Comment:
if going the C99-route, just write: `double coef_xy[n_coef_xy][2] = {0};`

Reply:
Unfortunately, this does not to work because C lacks broadcasting:
```
error: variable-sized object may not be initialized
```